sycl/source/detail/device_filter.cpp

```diff
@@ -124,6 +124,56 @@ void device_filter_list::addFilter(device_filter &Filter) {
   FilterList.push_back(Filter);
 }
 
+// Backend is compatible with the SYCL_DEVICE_FILTER in the following cases.
+// 1. Filter backend is '*' which means ANY backend.
+// 2. Filter backend match exactly with the given 'Backend'
+bool device_filter_list::backendCompatible(backend Backend) {
+  for (const device_filter &Filter : FilterList) {
+    backend FilterBackend = Filter.Backend;
+    if (FilterBackend == Backend || FilterBackend == backend::all)
+      return true;
+  }
+  return false;
+}
+
+bool device_filter_list::deviceTypeCompatible(info::device_type DeviceType) {
+  for (const device_filter &Filter : FilterList) {
+    info::device_type FilterDevType = Filter.DeviceType;
+    if (FilterDevType == DeviceType || FilterDevType == info::device_type::all)
+      return true;
+  }
+  return false;
+}
+
+bool device_filter_list::deviceNumberCompatible(int DeviceNum) {
+  for (const device_filter &Filter : FilterList) {
+    int FilterDevNum = Filter.DeviceNum;
+    if (!Filter.HasDeviceNum || FilterDevNum == DeviceNum)
+      return true;
+  }
+  return false;
+}
+
+bool device_filter_list::containsHost() {
+  for (const device_filter &Filter : FilterList) {
+    if (((Filter.Backend == backend::host) &&
+         (Filter.DeviceType == info::device_type::host ||
+          Filter.DeviceType == info::device_type::all)) ||
+        ((Filter.Backend == backend::all) &&
+         (Filter.DeviceType == info::device_type::host))) {
+      // SYCL RT never creates more than one HOST device.
+      // All device numbers other than 0 are rejected.
+      if (!Filter.HasDeviceNum || Filter.DeviceNum == 0)
+        return true;
+    }
+    if (Filter.Backend == backend::all &&
+        Filter.DeviceType == info::device_type::all &&
+        (!Filter.HasDeviceNum || Filter.DeviceNum == 0))
+      return true;
```

Comment:
Can we replace it with?:
```
if(Filter.Backend == backend::host || Filter.Backend == backend::all)
    if(Filter.DeviceType == info::device_type::all || Filter.DeviceType == info::device_type::host)
        if(!Filter.HasDeviceNum || Filter.DeviceNum == 0)
            return true;
```
BTW. Shouldn't we check if device type and backend are set similar to device num?

Reply:
Thanks. I accommodated your suggestion.
> BTW. Shouldn't we check if device type and backend are set similar to device num?
Not sure what you mean. Do you mean we should check if device type and backend are missing (unspecified)?
The device_filter constructor uses backend::all and device_type::all for missing entries.
So we do not need to check if they are missing.